### ISSUE TITLE
Handle special characters in paths correctly

### ISSUE THREAD

**Issue by marinier** [soartech/soar-language-server#24]
I'm tried Eclipse's generic LSP plugin. The way this works is you define a file association (e.g., to .soar and *.tcl files), and then you define a server that goes with that association (e.g., via an external launch). The server launches fine, but an exception gets thrown. The output below is for the "project" test example. It looks like it doesn't like the URI text, probably because it has a ":" in it (that didn't get encoded correctly?).

```
{
  "jsonrpc": "2.0",
  "id": "1",
  "method": "initialize",
  "params": {
    "processId": 26452,
    "rootPath": "/C:/Users/bob.marinier/eclipse workspaces/lsp-test/lsptest/",
    "rootUri": "file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/",
    "capabilities": {
      "workspace": {
        "applyEdit": true,
        "workspaceEdit": {
          "documentChanges": true,
          "resourceOperations": [
            "create",
            "delete",
            "rename"
          ],
          "failureHandling": "undo"
        },
        "symbol": {},
        "executeCommand": {
          "dynamicRegistration": true
        },
        "workspaceFolders": true
      },
      "textDocument": {
        "synchronization": {
          "willSave": true,
          "willSaveWaitUntil": true,
          "didSave": true
        },
        "completion": {
          "completionItem": {
            "snippetSupport": true
          }
        },
        "hover": {},
        "signatureHelp": {},
        "references": {},
        "documentHighlight": {},
        "documentSymbol": {
          "symbolKind": {
            "valueSet": [
              18,
              17,
              5,
              14,
              9,
              10,
              22,
              24,
              8,
              1,
              12,
              11,
              20,
              6,
              2,
              3,
              21,
              16,
              19,
              25,
              4,
              7,
              15,
              23,
              26,
              13
            ]
          },
          "hierarchicalDocumentSymbolSupport": true
        },
        "formatting": {},
        "rangeFormatting": {},
        "definition": {},
        "codeAction": {
          "codeActionLiteralSupport": {
            "codeActionKind": {
              "valueSet": [
                "quickfix",
                "refactor",
                "refactor.extract",
                "refactor.inline",
                "refactor.rewrite",
                "source",
                "source.organizeImports"
              ]
            }
          },
          "dynamicRegistration": true
        },
        "codeLens": {},
        "documentLink": {},
        "colorProvider": {},
        "rename": {}
      }
    },
    "clientName": "Eclipse IDE",
    "trace": "off"
  }
}
[INFO ] [2019-04-22 13:57:02,978] [pool-2-thread-1] [o.e.l.s.LanguageServer] Initializing server
[INFO ] [2019-04-22 13:57:02,981] [pool-2-thread-1] [c.s.s.SoarWorkspaceService] Setting workspace URI: file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/
[INFO ] [2019-04-22 13:57:02,981] [pool-2-thread-1] [c.s.s.SoarWorkspaceService] Processing entry points: workspace URI: file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/
{
  "jsonrpc": "2.0",
  "id": "1",
  "result": {
    "capabilities": {
      "textDocumentSync": 2,
      "hoverProvider": true,
      "completionProvider": {
        "resolveProvider": false,
        "triggerCharacters": [
          "$",
          "["
        ]
      },
      "signatureHelpProvider": {
        "triggerCharacters": [
          " "
        ]
      },
      "definitionProvider": true,
      "referencesProvider": true,
      "codeActionProvider": true,
      "foldingRangeProvider": true,
      "executeCommandProvider": {
        "commands": []
      }
    }
  }
}
Content-Length: 372

{"jsonrpc":"2.0","id":"1","result":{"capabilities":{"textDocumentSync":2,"hoverProvider":true,"completionProvider":{"resolveProvider":false,"triggerCharacters":["$","["]},"signatureHelpProvider":{"triggerCharacters":[" "]},"definitionProvider":true,"referencesProvider":true,"codeActionProvider":true,"foldingRangeProvider":true,"executeCommandProvider":{"commands":[]}}}}{
  "jsonrpc": "2.0",
  "method": "initialized",
  "params": {}
}
{
  "jsonrpc": "2.0",
  "method": "textDocument/didOpen",
  "params": {
    "textDocument": {
      "uri": "file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/load.soar",
      "languageId": "soar",
      "version": 1,
      "text": "# This is the root load.soar of a multi-file project.\r\n\r\nsource micro-ngs/load.soar\r\nsource productions.soar\r\n\r\n# Attempting to load this file should cause an error.\r\n#source missing-file.soar"
    }
  }
}
Apr 22, 2019 1:57:03 PM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleNotification
WARNING: Notification threw an exception: {
  "jsonrpc": "2.0",
  "method": "textDocument/didOpen",
  "params": {
    "textDocument": {
      "uri": "file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/load.soar",
      "languageId": "soar",
      "version": 1,
      "text": "# This is the root load.soar of a multi-file project.\r\n\r\nsource micro-ngs/load.soar\r\nsource productions.soar\r\n\r\n# Attempting to load this file should cause an error.\r\n#source missing-file.soar"
    }
  }
}
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:67)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:152)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:220)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:187)
	at org.eclipse.lsp4j.jsonrpc.Launcher$Builder.lambda$wrapMessageConsumer$0(Launcher.java:337)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:192)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:94)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:113)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:65)
	... 12 more
Caused by: java.lang.RuntimeException: java.net.URISyntaxException: Illegal character in path at index 37: file:///C:/Users/bob.marinier/eclipse workspaces/lsp-test/lsptest/load.soar
	at com.soartech.soarls.SoarDocumentService.uri(SoarDocumentService.java:693)
	at com.soartech.soarls.Documents.open(Documents.java:54)
	at com.soartech.soarls.SoarDocumentService.didOpen(SoarDocumentService.java:145)
	... 17 more
Caused by: java.net.URISyntaxException: Illegal character in path at index 37: file:///C:/Users/bob.marinier/eclipse workspaces/lsp-test/lsptest/load.soar
	at java.net.URI$Parser.fail(URI.java:2848)
	at java.net.URI$Parser.checkChars(URI.java:3021)
	at java.net.URI$Parser.parseHierarchical(URI.java:3105)
	at java.net.URI$Parser.parse(URI.java:3053)
	at java.net.URI.<init>(URI.java:588)
	at com.soartech.soarls.SoarDocumentService.uri(SoarDocumentService.java:691)
	... 19 more

[INFO ] [2019-04-22 13:57:04,543] [pool-1-thread-1] [o.j.u.c.SoarCommands] Registering custom commands from class org.jsoar.kernel.smem.SmemCommand$Provider
[INFO ] [2019-04-22 13:57:04,554] [pool-1-thread-1] [o.j.u.c.SoarCommands] Registering custom commands from class org.jsoar.kernel.epmem.EpmemCommand$Provider
[INFO ] [2019-04-22 13:57:04,555] [pool-1-thread-1] [o.j.u.c.SoarCommands] Registering custom commands from class org.jsoar.script.ScriptCommand$Provider
[INFO ] [2019-04-22 13:57:04,556] [pool-1-thread-1] [o.j.u.c.SoarCommands] Registering custom commands from class org.jsoar.kernel.wma.WMActivationCommand$Provider
[INFO ] [2019-04-22 13:57:04,556] [pool-1-thread-1] [o.j.u.c.SoarCommands] Registering custom commands from class org.jsoar.kernel.learning.rl.RLCommand$Provider
[INFO ] [2019-04-22 13:57:04,559] [pool-1-thread-1] [o.j.k.Agent] Current command interpreter is 'tcl' : 'class org.jsoar.tcl.SoarTclInterface'
[DEBUG] [2019-04-22 13:57:04,572] [pool-1-thread-1] [o.j.u.UrlTools] uri: jar:file:/C:/Users/bob.marinier/git/soar-language-server/build/install/soar-language-server/lib/jsoar-core-4.0-SNAPSHOT.jar!/org/jsoar/kernel/commands/aliases
[DEBUG] [2019-04-22 13:57:04,572] [pool-1-thread-1] [o.j.u.UrlTools] ssp1: file:/C:/Users/bob.marinier/git/soar-language-server/build/install/soar-language-server/lib/jsoar-core-4.0-SNAPSHOT.jar!/org/jsoar/kernel/commands/aliases
[DEBUG] [2019-04-22 13:57:04,572] [pool-1-thread-1] [o.j.u.UrlTools] ssp2: /C:/Users/bob.marinier/git/soar-language-server/build/install/soar-language-server/lib/jsoar-core-4.0-SNAPSHOT.jar!/org/jsoar/kernel/commands/aliases
[DEBUG] [2019-04-22 13:57:04,572] [pool-1-thread-1] [o.j.u.UrlTools] scheme: file
[DEBUG] [2019-04-22 13:57:04,572] [pool-1-thread-1] [o.j.u.UrlTools] normalzied ssp2: /C:/Users/bob.marinier/git/soar-language-server/build/install/soar-language-server/lib/jsoar-core-4.0-SNAPSHOT.jar!/org/jsoar/kernel/commands/aliases
[DEBUG] [2019-04-22 13:57:04,572] [pool-1-thread-1] [o.j.u.UrlTools] normalized: jar:file:/C:/Users/bob.marinier/git/soar-language-server/build/install/soar-language-server/lib/jsoar-core-4.0-SNAPSHOT.jar!/org/jsoar/kernel/commands/aliases
[INFO ] [2019-04-22 13:57:04,678] [pool-1-thread-1] [c.s.s.a.Analysis] Retrieved file for file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/micro-ngs/load.soar :: com.soartech.soarls.SoarFile@119c09a5
[INFO ] [2019-04-22 13:57:04,682] [pool-1-thread-1] [c.s.s.a.Analysis] Retrieved file for file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/micro-ngs/macros.tcl :: com.soartech.soarls.SoarFile@15db0395
[INFO ] [2019-04-22 13:57:04,703] [pool-1-thread-1] [c.s.s.a.Analysis] Retrieved file for file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/productions.soar :: com.soartech.soarls.SoarFile@25495d3e
[INFO ] [2019-04-22 13:57:04,854] [pool-1-thread-1] [c.s.s.a.Analysis] Completed analysis com.soartech.soarls.analysis.Analysis@1f3bafd7
[INFO ] [2019-04-22 13:57:04,857] [pool-1-thread-1] [o.j.k.Agent] Current command interpreter is 'none'
[INFO ] [2019-04-22 13:57:04,858] [pool-1-thread-1] [o.j.k.Agent] Agent 'JSoar Agent 1' disposed.
{
  "jsonrpc": "2.0",
  "method": "textDocument/publishDiagnostics",
  "params": {
    "uri": "file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/productions.soar",
    "diagnostics": []
  }
}
Content-Length: 181

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/productions.soar","diagnostics":[]}}{
  "jsonrpc": "2.0",
  "method": "textDocument/publishDiagnostics",
  "params": {
    "uri": "file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/micro-ngs/load.soar",
    "diagnostics": []
  }
}
{
  "jsonrpc": "2.0",
  "method": "textDocument/publishDiagnostics",
  "params": {
    "uri": "file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/load.soar",
    "diagnostics": []
  }
}
{
  "jsonrpc": "2.0",
  "method": "textDocument/publishDiagnostics",
  "params": {
    "uri": "file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/micro-ngs/macros.tcl",
    "diagnostics": []
  }
}
Content-Length: 184

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/micro-ngs/load.soar","diagnostics":[]}}Content-Length: 174

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/load.soar","diagnostics":[]}}Content-Length: 185

{"jsonrpc":"2.0","method":"textDocument/publishDiagnostics","params":{"uri":"file:///C:/Users/bob.marinier/eclipse%20workspaces/lsp-test/lsptest/micro-ngs/macros.tcl","diagnostics":[]}}
```

**Comment by adeschamps** [comment]
The `URISyntaxException` calls out an illegal character at index 37, which is the space in `eclipse workspace`.

All URIs are generated by the `SoarDocumentService.uri()`. I don't know exactly _how_ this should be fixed, but I think that's _where_ it should be fixed.

**Comment by marinier** [comment]
The issue here seems to be that we're encoding URLs incorrectly. `URLEncoder` is meant for form data, not the base URLs themselves. See the discussion here:

https://stackoverflow.com/questions/4737841/urlencoder-not-able-to-translate-space-character

What's happening, then, is that `LanguageServerTestFixture.fileId()` is completely mangling the URIs (escaping things that should not be escaped, like all the slashes). To make a URI out of this later, `SoarDocumentService.uri()` decodes this, but that also restores spaces, which are not valid in URIs -- hence the exception.

There seem to be two parts to the fix:

(1) `SoarDocumentService.uri()` should use `new URL(uriString).toURI()` -- this handles the escaping properly from a string form.

(2) `LanguageServerTestFixture.fileId()` should only replace the special characters that need replacing (like spaces), not slashes. Unfortunately, there doesn't seem to be a great out-of-the-box way to do this. I tried the `new URL(uriString).toURI()` mentioned above, but then a number of other tests fail (e.g., it looks like `Location` objects don't have valid URIs?). Anyway, none of our tests use paths that contain special characters anyway, so if we were doing this correctly (and URIs everywhere were valid), we wouldn't actually expect this to do anything.

**Comment by marinier** [comment]
I added SpecialPathCharsTest (currently ignored) that directly tests the issue. This should allow us to remove the encoding step from the test harness, which was trying to fake this.

**Comment by adeschamps** [comment]
Closed in #54.
